docs/specifications/lease.rst

```diff
@@ -0,0 +1,45 @@
+.. -*- coding: utf-8 -*-
+
+.. _share leases:
+
+Share Leases
+============
+
+A lease is a marker attached to a share indicating that some client has asked for that share to be retained for some amount of time.
+The intent is to allow clients and servers to collaborate to determine which data should still be retained and which can be discarded to reclaim storage space.
+Zero or more leases may be attached to any particular share.
+
+Renewal Secrets
+---------------
+
+Each lease is uniquely identified by its **renewal secret**.
+This is a 32 byte string which can be used to extend the validity period of that lease.
+
+To a storage server a renewal secret is an opaque value which is only ever compared to other renewal secrets to determine equality.
+
+Storage clients will typically want to follow a scheme to deterministically derive the renewal secret for a particular share from information the client already holds about that share.
+This allows a client to maintain and renew single long-lived lease without maintaining additional local state.
+
+The scheme in use in Tahoe-LAFS as of 1.16.0 is as follows.
+
+* The **netstring encoding** of a byte string is the concatenation of:
+
+  * the ascii encoding of the base 10 representation of the length of the string
+  * ``":"``
+  * the string itself
+  * ``","``
+
+* The **sha256d digest** is the **sha256 digest** of the **sha256 digest** of a string.
```

Comment:
This whole section would be a lot clearer if there was also some code included to demonstrate (with type annotations).

Reply:
something like this?

```
"""                                                                                                                                                                                           
This is a reference implementation of the lease renewal secret derivation                                                                                                                     
protocol in use by Tahoe-LAFS clients as of 1.16.0.                                                                                                                                           
"""

from base64 import (
    b32encode,
)

from allmydata.util.hashutil import (
    tagged_hash,
    tagged_pair_hash,
)


def derive_renewal_secret(lease_secret: bytes, storage_index: bytes, tubid: bytes) -> bytes:
    assert len(lease_secret) == 32
    assert len(storage_index) == 20
    assert len(tubid) == 20

    bucket_renewal_tag = b"allmydata_bucket_renewal_secret_v1"
    file_renewal_tag = b"allmydata_file_renewal_secret_v1"
    client_renewal_tag = b"allmydata_client_renewal_secret_v1"

    client_renewal_secret = tagged_hash(lease_secret, client_renewal_tag)
    file_renewal_secret = tagged_pair_hash(
        file_renewal_tag,
        client_renewal_secret,
        storage_index,
    )
    peer_id = b32encode(tubid).lower().strip(b"=")

    return tagged_pair_hash(bucket_renewal_tag, file_renewal_secret, peer_id)

print(derive_renewal_secret(
    b"lease secretxxxxxxxxxxxxxxxxxxxx",
    b"storage indexxxxxxxx",
    b"tub idxxxxxxxxxxxxxx",
))
```

I'm not really sure where to put it, though, or what kind of quality control to try to apply to it ... I suppose I could try to observe a few derived renewal secrets from the actual client software and then capture them as test vectors ... But that code isn't exactly easy to invoke.

Reply:
Yeah that is what I was thinking. You could just put it after the prose section.

I was able to validate that implementation by looking at existing code, other than the logic that calculates `peer_id`. But... that logic doesn't match the prose spec anyway, since it's supposed to be based on x509 certificate, not tub ID? So if you fix that line I think it's fine.

Reply:
The tubid is the sha1 hash of the x509 certificate.  So I think it works out correctly the way I implemented it above.  It's true that "tubid" is a Foolscap concept and the code that computes it is in Foolscap.

I didn't feel like grabbing `cryptography` and making the certificate an input to this function.  The tubid is currently quite convenient to find (it's in every storage fURL) and that won't *immediately* change even after we drop Foolscap (because Tahoe will have to be capable of accepting fURLs for a long, long time).

So I'd like to leave this as-is is for now but I would be happy to see it revisited in the near future to scrub all traces of Foolscap from the spec.